python/qemu/ot/rom/image.py

```
@@ -343,6 +320,55 @@ def _save_hex(self, rfp: BinaryIO) -> None:
                             for pa in range(len(scrwords))).encode())
         rfp.write(b'\n')
 
+    def _load_vmem(self, rfp: BinaryIO, size: Optional[int] = None) -> None:
+        words: list[int] = []  # 64-bit values
+        scrambled: Optional[bool] = None
+        next_addr = 0
+        for lpos, line in enumerate(rfp.readlines(), start=1):
+            line = line.strip()
+            if not line.startswith(b'@'):
+                continue
+            parts = line[1:].split(b' ')
+            address_str = parts[0]
+            data_str = parts[1:]
+            scrambled_data = all(len(d) == 10 for d in data_str)
+            plain_data = all(len(d) == 8 for d in data_str)
+            if not scrambled_data ^ plain_data:
+                raise ValueError(f'Unknown VMEM format @ {lpos}')
+            if scrambled is None:
+                scrambled = scrambled_data
+                self._log.info('Identified %s as %s VMEM format',
+                               rfp.name or '?',
+                               'scrambled' if scrambled else 'plain')
+            elif scrambled != scrambled_data:
+                raise ValueError(f'Incoherent VMEM format @ {lpos}')
+            try:
+                address = int(address_str, 16)
+                words.extend(int(d, 16) for d in data_str)
+            except (TypeError, ValueError) as exc:
+                raise ValueError(f'Invalid data in VMEM format @ '
+                                 f'{lpos}') from exc
+            if address != next_addr:
+                raise ValueError(f'Incoherent next address @ {lpos}')
+            next_addr += len(data_str)
+        if scrambled is None:
+            self._log.error('No valid data found in VMEM file')
+            return
+        if scrambled:
+            self._handle_scrambled_data(words)
```

Should this maybe error if `size` is provided and it doesn't match the length of the parsed words (as bytes), as right now `size` is entirely discarded for the scrambled case? Or maybe just raise an error if `size` is provided for the scrambled case?

---

ok

---

ok, adding the check